### ISSUE TITLE
Redefinition of an Unknown creates two instances for hashing

### ISSUE THREAD

**Issue by jpfairbanks** [JuliaDiffEq/ModelingToolkit.jl#128]
If you redefine a variable with the same name, you get two instances from the perspective of hashing, but only variables are singleton from the perspective of `==`

MWE:
```julia

julia> @Unknown t x(t)
(Unknown(t), Unknown(x))

julia> [t, x.dependents[1]]
2-element Array{Variable,1}:
 Unknown(t)
 Unknown(t)

julia> Set([t, x.dependents[1]])
Set(Variable[Unknown(t)])

julia> @Unknown t
(Unknown(t),)

julia> Set([t, x.dependents[1]])
Set(Variable[Unknown(t), Unknown(t)])

julia> t == x.dependents[1]
true
```

(edit: accidentally hit send too soon)

**Comment by jpfairbanks** [comment]
I don't know if this is a problem, but it was weird when I saw two copies of Unknown(t) in the same set.

**Comment by ChrisRackauckas** [comment]
Probably needs some `==` overload?

**Comment by HarrisonGrodin** [comment]
This is on an older version; it seems like now, we actually have the opposite issue:

```julia
julia> @variables t x(t)
(Variable(), Variable(Variable()))

julia> t2 = t
Variable()

julia> @variables t
(Variable(),)

julia> Set([t, t2])
Set(Operation[Variable()])

julia> hash(t)
0xf44edc67490f062f

julia> hash(t2)
0x2d6b986e69de437f
```

**Comment by jpfairbanks** [comment]
re: version, I did "] dev PackageName" and assumed I was on the newest version, but I guess that's not a safe assumption, I'll upgrade with `git checkout master` now.

So based on this output:
```julia

julia> Set([t, t2])
Set(Operation[Variable()])

julia> hash(t)
0xf44edc67490f062f

julia> hash(t2)
0x2d6b986e69de437f
```

It appears that Set correctly identifies them as the same variable, even though they have different hashes. That sounds right to me, although is there any reason for distinguishing between the two "copies" of `t` in ModelingToolkit?

Also, I like the old printing style better, because it had the variable names in there that seems really useful.

**Comment by HarrisonGrodin** [comment]
It's actually kind of backwards - now, we expect all instances of `Variable` to be distinct, with names being treated as purely visual. (We should have `!isequal(Variable(:x), Variable(:x))`.) However, it seems like while hashing reflects this, `isequal`/etc. haven't been updated. See #89 for more info.

Also, the old printing style is definitely better; I'm not sure what happened here. I opened #129 in regard to this specifically.

**Comment by jpfairbanks** [comment]
Closed in favor of #89 and #129
